### PR TITLE
fused linear and selective recompute

### PR DESCRIPTION
**Pull request by FeixLiu**
Part of pr https://github.com/PaddlePaddle/FleetX/pull/613, only contains the fused linear part and selective recompute part. 

### Loss compare for pure data parallel training (dp = 8)
tensor_fusion + fused_linear + selective_recompute
image

#### 这个性能有fused_linear
| baseline speed | optimize speed | gain |
|:--:|:--:|:--:|
|181012|240203|+32.7%|

### Loss compare for hybrid parallel training (dp=mp=pp=2)
fused_linear + selective_recompute

image

| baseline speed | optimize speed | gain |
|:--:|:--:|:--:|
|19072|21241|+11.4%|

### CODE REVIEW

File: examples/gpt/hybrid_parallel/README.md [hunk 1]
```diff
@@ -95,6 +95,7 @@ GPT训练默认使用AdamW优化器以及cosine 学习率衰减，这里通过
   num_train_epochs: 1
   seed: 1024
   use_recompute: False
+  recompute_granularity:
```
Comment (sneaxiy) [inline examples/gpt/hybrid_parallel/README.md:98]: 这个要填`full`吗？

File: examples/gpt/tools.py [hunk 2]
```diff
@@ -84,6 +92,22 @@ def add_dict(config, k, v):
 
     args.test_iters = args.eval_iters * 10
 
+    if args.fused_linear:
+        assert is_fused_matmul_bias_supported(), \
```
Comment (sneaxiy) [inline examples/gpt/tools.py]: 是直接报错，还是退化比较合适？

File: examples/gpt/hybrid_parallel/README.md (hunk 1 above)
Comment (FeixLiu) [reply examples/gpt/hybrid_parallel/README.md:98]: 不需要，这个recompute是false，空着就行，在backend收到的是一个None

File: examples/gpt/tools.py (hunk 2 above)
Comment (FeixLiu) [reply examples/gpt/tools.py]: 退化吧，已update

**Review by ForFishes — APPROVED** [review]

**Pull request merged**